src/tools/illink/test/Mono.Linker.Tests.Cases/DataFlow/FeatureCheckDataFlow.cs

```diff
@@ -0,0 +1,999 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using ILLink.RoslynAnalyzer;
+using Mono.Linker.Tests.Cases.Expectations.Assertions;
+using Mono.Linker.Tests.Cases.Expectations.Helpers;
+
+namespace Mono.Linker.Tests.Cases.DataFlow
+{
+	[SkipKeptItemsValidation]
+	[ExpectedNoWarnings]
+	public class FeatureCheckDataFlow
+	{
+		public static void Main ()
+		{
+			CallFeatureUnguarded.Test ();
+			CallFeatureGuarded.Test ();
+			FeatureCheckBooleanExpressions.Test ();
+			SupportedFeatureChecks.Test ();
+			FeatureCheckCombinations.Test ();
+			GuardedPatterns.Test ();
+			ExceptionalDataFlow.Test ();
+		}
+
+		class CallFeatureUnguarded
+		{
+			[ExpectedWarning ("IL3050", nameof (RequiresDynamicCode), ProducedBy = Tool.Analyzer | Tool.NativeAot)]
+			static void Unguarded ()
+			{
+				RequiresDynamicCode ();
```

Comment:
For completeness I would add RUC and RAF to this and maybe the if test below. As a way to validate that the feature check for RDC doesn't accidentally also turn off RUC/RAF warnings.
Not needed for all the combinations, just a couple.